jerry-core/ecma/operations/ecma-promise-object.c

```diff
@@ -0,0 +1,711 @@
+                               const ecma_value_t this, /**< this_arg of the function */
+                               const ecma_value_t argv[], /**< argument list */
+                               const ecma_length_t argc) /**< argument number */
+{
+  JERRY_UNUSED (this);
+  JERRY_UNUSED (argc);
+
+  ecma_string_t *str_0 = ecma_new_ecma_string_from_uint32 (0);
+  ecma_string_t *str_1 = ecma_new_ecma_string_from_uint32 (1);
+  ecma_string_t *str_2 = ecma_new_ecma_string_from_uint32 (2);
+  /* 2. String '0' indicates [[Capability]] of the function object. */
+  ecma_value_t capability = ecma_op_object_get (ecma_get_object_from_value (function), str_0);
+  /* 3. String '1' indicates [[Resolve]] of the capability. */
+  ecma_value_t resolve = ecma_op_object_get (ecma_get_object_from_value (capability), str_1);
+
+  if (resolve != ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED))
+  {
+    ecma_free_value (resolve);
+    ecma_free_value (capability);
+    ecma_deref_ecma_string (str_0);
+    ecma_deref_ecma_string (str_1);
+    ecma_deref_ecma_string (str_2);
+
+    return ecma_raise_type_error (ECMA_ERR_MSG ("'resolve' function should be undefined."));
+  }
+  /* 4. String '2' indicates [[Reject]] of the capability. */
+  ecma_value_t reject = ecma_op_object_get (ecma_get_object_from_value (capability), str_2);
+
+  if (reject != ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED))
+  {
+    ecma_free_value (reject);
+    ecma_free_value (capability);
+    ecma_deref_ecma_string (str_0);
+    ecma_deref_ecma_string (str_1);
+    ecma_deref_ecma_string (str_2);
+
+    return ecma_raise_type_error (ECMA_ERR_MSG ("'reject' function should be undefined."));
+  }
+
+  /* 5. */
+  ecma_op_object_put (ecma_get_object_from_value (capability),
+                      str_1,
+                      argv[0],
+                      false);
+  /* 6. */
+  ecma_op_object_put (ecma_get_object_from_value (capability),
+                      str_2,
+                      argv[1],
+                      false);
+
+  ecma_free_value (capability);
+  ecma_deref_ecma_string (str_0);
+  ecma_deref_ecma_string (str_1);
+  ecma_deref_ecma_string (str_2);
+
+  return ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+} /* ecma_promise_executor_handler */
+
+/**
+ * Create a PromiseCreateResovingFucntions.
+ *
+ * See also: ES2015 25.4.1.3
+ *
+ * @return pointer to the resolving functions
+ */
+ecma_promise_resolving_functions_t *
+ecma_promise_create_resolving_functions (ecma_object_t *object_p) /**< the promise object */
+{
+  /* 1. */
+  ecma_value_t already_resolved = ecma_op_create_boolean_object (ecma_make_boolean_value (false));
+
+  ecma_string_t *str_0 = ecma_new_ecma_string_from_uint32 (0);
+  ecma_string_t *str_1 = ecma_new_ecma_string_from_uint32 (1);
+  /* 2. */
+  ecma_object_t *resolve_p;
+  resolve_p = ecma_op_create_external_function_object ((ecma_external_pointer_t) ecma_promise_resolve_handler);
```

Comment:
1). resolve_handler and reject_handler are not only used internally, in jobqueue's `ecma_process_promise_resolve_thenable_job`. it will be used as args of thenable's then function. And the then function there, could be implemented by developer's binding code or js code. So I think it is needed to create JS function for `ecma_promise_resolve_handler` and `ecma_promise_reject_handler`.

2). you're right in your [last comment](https://github.com/jerryscript-project/jerryscript/pull/1695#discussion_r109372202), because reject and resolve will be passed to the user's custom `then` function,  we cannot define a normal string property for `function_p`. I will use LIT_INTERNAL_MAGIC_STRING_XXX as the property name, so that outer user can not get access to the property. 

3).While the creation of `ecma_promise_executor_handler` is not necessary, and I will change it.